Introduce a gate/check GHA job

webknjaz opened the pull request:
This adds a GHA job that reliably determines if all the required dependencies have succeeded or not.

It also allows to reduce the list of required branch protection CI statuses to just one — `check`. This reduces the maintenance burden by a lot and have been battle-tested across a small bunch of projects in its action form and in-house implementations of other people.

This action is now in use in aiohttp (and other aio-libs projects), CherryPy, some of the Ansible repositories, pip-tools, all of the jaraco's projects (like `setuptools`, `importlib_metadata`), some of the hynek's projects (like `attrs`, `structlog`), some PyCQA, PyCA, PyPA and pytest projects, a few AWS Labs projects. Admittedly, I maintain a few of these but it seems to address some of the pain folks have: https://github.com/jaraco/skeleton/pull/55#issuecomment-1106638475.

I saw a few attempts to reduce the maintenance burden for GHA and figured
this might be useful for CPython too, which is why I'm submitting this patch.
Looking forward to hearing what you think!

The story behind this is explained in more detail at https://github.com/marketplace/actions/alls-green#why.

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -116,6 +116,7 @@ jobs:
     runs-on: windows-latest
     needs: check_source
     if: needs.check_source.outputs.run_tests == 'true'
+    continue-on-error: true
```
By default, all the jobs listed in the `check` job's `needs` section are required to succeed in terms of branch protection. This setting emulates having the job non-required (in combination with the check job below being the only thing required in branch protection).

webknjaz replied:
This has been moved to the action call. Resolving.

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -317,3 +320,44 @@ jobs:
       run: xvfb-run make buildbottest TESTOPTS="-j4 -uall,-cpu"
+
+  check:  # This job does nothing and is only used for the branch protection
+    if: always()
```
This is mandatory because otherwise it'll get a `skipped` status in some cases and the branch protection sees that as `skipped == success` which is undesired.

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -317,3 +320,44 @@ jobs:
+    - name: Decide whether the needed jobs succeeded or failed
+      uses: re-actors/alls-green@198badcb65a1a44528f27d5da555c4be9f12eac6
+      with:
+        allowed-skips: >-
```
This setting tells the action that if the listed jobs (comma-separated) got skipped, that's okay. But the jobs are allowed to be skipped based on the same condition they've got set. If the first job requests test runs, this list will compute as empty and none of the jobs will be allowed to be skipped (meaning that skips would turn into failures).

hugovk replied:
Do I understand this right?

- If `check-docs` starts, it must pass; if they skip, that's fine.

- If `check_generated_files` or any of the other jobs in its group start, they must all pass; if they all skip, that's fine.

- If `test_hypothesis` starts, it must pass; if they skip, that's fine.

webknjaz replied:
@hugovk almost: this is paired with the setting above :point_up: — essentially, if something is allowed to be skipped, and it is skipped, the action will treat it as non-failure.
But if something is allowed to be skipped and is not skipped, whether its failure is to be treated as such is controlled by the `allowed-failures` input.

Also, not they are not grouped inside the action, the templating here just produces a comma-separated list of things and the action doesn't know it was in some group externally. So yes, such a relation kinda exists, but I would say that it's indirect.

And since there's no “groups”, if any of the jobs that are allowed to be skipped run, each individual job is evaluated separately from the “group”, based on the fact of failures being allowed or not.

N.B. Whether a job is allowed to fail is also contributed by the `continue-on-error` setting, which is mostly useful for controlling individual matrix jobs. If a job was allowed to fail prior to this one, the action will not “see” if it was red and treat it as green.

webknjaz replied:
> * If `test_hypothesis` starts, it must pass; if they skip, that's fine.

With the context I added earlier, seeing that `test_hypothesis` is listed in `allowed-failures` unconditionally, it will never cause failures. This reflects the current branch protection configuration as seen in the Checks widget in PRs.

ambv commented on .github/workflows/build.yml:
```diff
@@ -318,3 +321,38 @@ jobs:
+
+    steps:
+    - name: Decide whether the needed jobs succeeded or failed
+      uses: re-actors/alls-green@13b4244b312e8a314951e03958a2f91519a6a3c9
```
What's the reason this is pinned to v1.2.1 instead of using the release name? Also, I looked at the repo of the action and the checks are red there, is that expected?

webknjaz replied:
> What's the reason this is pinned to v1.2.1 instead of using the release name?

No particular reason from my side but when I contributed this to other projects, some wanted to have commit sha pinning. Dependabot is able to bump such commit hash based pins so it's usually considered the best of both worlds.

For myself, I usually do this:
```suggestion
      uses: re-actors/alls-green@release/v1
```
But I appreciate that some people want better traceability for non-GH-official actions. It's up to you really.

If you choose to go for the `release/v1` branch, it represents stable API. Note that Dependabot will not be updating the reference in this case.

> Also, I looked at the repo of the action and the checks are red there, is that expected?

Yes, kinda. It is necessary for internal smoke-testing. Some of the jobs are red but the check job isn't. The overall workflow status may be red but it's not necessary a reflection of the gate check, it's just a quirk of how GitHub represents statuses — it doesn't know that we actually care about the outcome of one single job which decides internally whether the dependencies are allowed to fail or be skipped, and if any of the workflow jobs is red, it marks the whole thing as red. The branch protection is a disconnected concept and is not applied in all possible contexts, which is why GH can't make assumptions automatically, and it doesn't provide a way to declare what's important on the GHA declaration level.
It's weird but we can't do anything about it, unless GitHub decides to make updates...

ambv replied:
Alright, all this makes sense to me.

ambv commented on .github/workflows/build.yml:
```diff
@@ -339,3 +342,38 @@ jobs:
     - name: Tests
       run: xvfb-run make buildbottest TESTOPTS="-j4 -uall,-cpu"
+
+  check:  # This job does nothing and is only used for the branch protection
```
"All required checks pass"

ambv replied:
Job ID: `s/check/all-required-green/`

webknjaz replied:
Patched. Marking the conversation as resolved.

webknjaz committed:
✅ Introduce a gate/check GHA job

This adds a GHA job that reliably determines if all the required
dependencies have succeeded or not.

It also allows to reduce the list of required branch protection CI
statuses to just one — `check`. This reduces the maintenance burden
by a lot and have been battle-tested across a small bunch of projects
in its action form and in-house implementations of other people.

This action is now in use in aiohttp (and other aio-libs projects),
CherryPy, conda, coveragepy, Open edX, Towncrier some of the Ansible
repositories, pip-tools, all of the jaraco's projects (like
`setuptools`, `importlib_metadata`), some of hynek's projects (like
`attrs`, `structlog`), some PyCQA, PyCA, PyPA and pytest projects, a
few AWS Labs projects. Admittedly, I maintain a few of these but it
seems to address some of the pain folks have:
https://github.com/jaraco/skeleton/pull/55#issuecomment-1106638475.

I figured, this might be useful for CPython too, which is why I'm
submitting this patch.

The story behind this is explained in more detail at
https://github.com/marketplace/actions/alls-green#why.

Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

AA-Turner commented on .github/workflows/reusable-docs.yml:
```diff
@@ -1,29 +1,8 @@
-name: Docs
+name: 📝
```
```suggestion
name: Docs
```

AA-Turner replied:
(I find the intented meaning of emoji harder to recognise at a glance, especially when I'm switching contexts -- words are much more explicit!)

webknjaz replied:
This is specifically done to make the UI displayed things shorter. So I'll wait for other people's feedbacks to see if it makes sense to revert to the previous name.
The thing is that with reusable workflows, the strings will be longer. So I kept the internal job name, but not the outer layer things.

hugovk replied:
Let's keep "Docs", it's short enough already.

By the way, we usually don't force push to https://github.com/python/ repos, it makes it a bit harder to review, and everything gets squash merged anyway. Thanks!

PS it was great to meet you at PyCon!

AlexWaygood replied:
I agree with Adam and Hugo, I strongly prefer words over emojis, especially when they're short! :)

webknjaz replied:
Please confirm that you are fine seeing a multi-segment thing like `Tests / Docs / Docs` in the UI with duplicate parts. I'm not sure you fully understand the implication of how this influences the check names.

webknjaz replied:
> By the way, we usually don't force push to https://github.com/python/ repos, it makes it a bit harder to review, and everything gets squash merged anyway.

Yes, sometimes it makes it easier but I'll need to rebase after the separate PR is merged anyway. Plus it makes me easier to smuggle a nice commit message into the resulting thing.

AlexWaygood replied:
> Please confirm that you are fine seeing a multi-segment thing like `Tests / Docs / Docs` in the UI. I'm not sure you fully understand the implication of how this influences the check names.

I think I'd be fine with having `Tests / Docs / Docs` in the list of checks at the bottom of the PR if it means we can avoid this, when you click through to view the results of the checks:

![6E362A53-9461-4E8B-89B8-8037B1770EE9](https://user-images.githubusercontent.com/66076021/235082176-10072374-650e-4712-9453-4e47be48d0f5.jpeg)

webknjaz replied:
Oh, so you're talking about a different view. That's fair. I'd like to hear from others and @llanga but now I at least understand where this is coming from...

hugovk replied:
I'm fine with:

* Tests / Docs / Docs
* Tests / Docs / Docs (Oldest Sphinx)
* Tests / Docs / Doctest

We can maybe rename that final part later, but for now I prefer to keep words.

(And it's @ambv not `@llanga` on GitHub :)

webknjaz replied:
Oh, thanks for correcting me 🤦‍♂️

hugovk replied:
> Yes, sometimes it makes it easier but I'll need to rebase after the separate PR is merged anyway. Plus it makes me easier to smuggle a nice commit message into the resulting thing.

The nice commit message is lost when we squash merge, so don't worry about it. And we can merge `main` into here when the other PR is merged, no need to rebase.

Shall we mark this PR as draft and continue on the other one?

webknjaz replied:
> We can maybe rename that final part later, but for now I prefer to keep words.

Note that this specific line refers to the middle part, not the final one.

webknjaz replied:
> The nice commit message is lost when we squash merge, so don't worry about it. And we can merge main into here when the other PR is merged, no need to rebase.

Well, it's a habit I cultivate and have to keep doing because habits don't work if you apply them selectively. This is also the reason I don't like squashes since they ruin history in certain ways.

> Shall we mark this PR as draft and continue on the other one?

Yeah, I was considering this. I've been basically waiting for Łukasz to finish his journey back home to resume the work on this. I'll be leaving SLC in about 12 hours myself and probably won't get to the laptop until the Saturday evening (Prague time).

webknjaz commented on .github/workflows/build.yml:
```diff
@@ -519,3 +519,51 @@ jobs:
+    name: ✅ All required checks pass
+    if: always()
+
+    needs:
```
TODO: add hypothesis here too

hugovk commented on .github/workflows/build.yml:
```diff
@@ -519,3 +519,60 @@ jobs:
       run: xvfb-run make buildbottest TESTOPTS="-j4 -uall,-cpu"
+
+  all-required-green:  # This job does nothing and is only used for the branch protection
+    name: ✅ All required checks pass
```
Let's remove the emoji, it's a little odd to see a double green (and we could sometimes see a "❌✅ All required checks pass"):

<img width="333" alt="image" src="https://github.com/python/cpython/assets/1324225/7d165a39-b339-4f4c-a177-1ab6ddd2cca2">


```suggestion
    name: All required checks pass
```

webknjaz replied:
Done!

hugovk commented on .github/workflows/build.yml:
```diff
@@ -519,3 +519,60 @@ jobs:
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Decide whether the needed jobs succeeded or failed
```
```suggestion
    - name: Check whether the needed jobs succeeded or failed
```

webknjaz replied:
Done!

hugovk reviewed (approved):
Thanks! Let's give @ambv some time to check before merging.
